tests/net/lib/coap/src/main.c

```diff
@@ -1855,4 +1855,87 @@ ZTEST(coap, test_age_is_newer)
 		     "Max age should be marked as newer");
 }
 
+struct test_coap_request {
+	uint16_t id;
+	uint8_t token[COAP_TOKEN_MAX_LEN];
+	uint8_t tkl;
+	enum coap_msgtype type;
+	struct coap_reply *match;
+};
+
+static int reply_cb(const struct coap_packet *response,
+		    struct coap_reply *reply,
+		    const struct sockaddr *from)
+{
+	return 0;
+}
+
+ZTEST(coap, test_response_matching)
+{
+	struct coap_reply matches[] = {
+		{ .id = 100, .reply = reply_cb },
+		{ .id = 101, .token = { 1, 2, 3, 4 }, .tkl = 4, .reply = reply_cb },
+	};
+	struct test_coap_request test_responses[] = {
+		/* Piggybacked ACK, no token */
+		{ .id = 100, .type = COAP_TYPE_ACK, .match = &matches[0] },
+		/* Piggybacked ACK, matching token */
+		{ .id = 101, .type = COAP_TYPE_ACK, .match = &matches[1],
+		  .token = { 1, 2, 3, 4 }, .tkl = 4  },
+		/* Piggybacked ACK, token mismatch */
+		{ .id = 101, .type = COAP_TYPE_ACK, .match = NULL,
+		  .token = { 1, 2, 3, 3 }, .tkl = 4 },
+		/* Piggybacked ACK, token mismatch 2 */
+		{ .id = 100, .type = COAP_TYPE_ACK, .match = NULL,
+		  .token = { 1, 2, 3, 4 }, .tkl = 4 },
+		/* Piggybacked ACK, token mismatch 3 */
+		{ .id = 101, .type = COAP_TYPE_ACK, .match = NULL },
+		/* Piggybacked ACK, id mismatch */
+		{ .id = 102, .type = COAP_TYPE_ACK, .match = NULL,
+		  .token = { 1, 2, 3, 4 }, .tkl = 4 },
+		/* Separate reply, no token */
+		{ .id = 100, .type = COAP_TYPE_CON, .match = NULL },
```

Comment:
For separate replies, there is no relation in the MIDs!
Using different MIDs for this test, e.g. 1000, would show that.